### PR TITLE
Add Picasa importer

### PR DESCRIPTION
**Pull request by kalman**
Current status:
  - slurp works, and it can request its own refresh token, and convert
    those refresh tokens into access tokens as needed.
  - find-photos is just a skeleton, it's never even been run.

### CODE REVIEW

**Comment by kalman** [comment]
This is a work in progress @aboodman 

Modulo review, the slurper should be submittable. In fact if you review that really quickly I could even submit it tomorrow morning.

I would take out the find-photos client, though.

**Comment by kalman** [comment]
Removed find-photos, PTAL.

File: samples/js/picasa/slurp/src/main.js [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+      progFn(more);
+    }
+    out.push(...more);
+  }
+  return out;
+}
+
+function concatApiEntries<T>(api: Api<T>[]): T[] {
+  const entries = api.map(a => a.feed.entry);
+  return entries.length === 0 ? [] : entries[0].concat(...entries.slice(1));
```
Comment (arv) [inline samples/js/picasa/slurp/src/main.js]: no need for `...` with concat. concat is magic.

File: samples/js/picasa/slurp/src/oauth.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
+// @flow
+
+// Copyright 2016 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+import {createServer as createHttpServer} from 'http';
+import {OAuth2} from 'oauth';
```
Comment (arv) [inline samples/js/picasa/slurp/src/oauth.js]: how generic is this? Can it be shared by other "slurpers"?

File: samples/js/picasa/slurp/src/oauth.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        res(accessToken);
+      }
+    });
+  });
+}
+
+function getAuthCodeViaURL(clientId: string, clientSecret: string)
+    : Promise<[string /* auth code */, string /* authorize URL */]> {
+  return new Promise((res, rej) => {
+    const server = createHttpServer((request, response) => {
```
Comment (arv) [inline samples/js/picasa/slurp/src/oauth.js]: what is going on here?

File: samples/js/picasa/slurp/src/main.js [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+  console.log(`\nSlurped ${numPhotosSoFar} photos`);
+
+  const albums = concatApiEntries((albumsArray: Albums[]));
+  const faces = concatApiEntries((facesArray: Faces[]));
+
+  // Note: albums and faces contain a lot of duplicate data. faces basically contains everything in
+  // albums, plus face data. This is pretty wasteful and causes commit to take twice as long as
+  // necessary. Consider doing something about that.
+
+  const j = (v: any) => jsonToNoms(v);
```
Comment (kalman) [inline samples/js/picasa/slurp/src/main.js]: Flow

File: samples/js/picasa/slurp/src/oauth.js (hunk 2 above)
Comment (kalman) [inline samples/js/picasa/slurp/src/oauth.js]: Anything which needs refresh tokens yes, but nothing else does yet. We can cross that bridge.

File: samples/js/picasa/slurp/src/main.js [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+  Api,
+  Albums,
+  Faces,
+  User,
+} from './types.js';
+import {getRefreshToken, getAccessTokenFromRefreshToken} from './oauth.js';
+
+const args = argv
+  .usage(
+    'Parses photo metadata out of Picasa API\n\n' +
```
Comment (kalman) [inline samples/js/picasa/slurp/src/main.js]: done

File: samples/js/picasa/slurp/src/main.js [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+    'must be re-run with that refresh token. Hold on to it.\n\n' +
+    'In other words:\n' +
+    ' node . --access-token <dest-dataset>\n' +
+    'or\n' +
+    ' node . --client-id=<id> --client-secret=<secret> [--refresh-token=<token> <dest-dataset>]\n')
+  .option('access_token', {
+    describe: 'OAuth2 access token (these are short lived and should only be used for testing)',
+    type: 'string',
+  })
+  .option('client_id', {
```
Comment (kalman) [inline samples/js/picasa/slurp/src/main.js]: @aboodman could you do this in a follow up? I don't have time (literally only have 30 seconds)

File: samples/js/picasa/slurp/src/main.js [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+    if (Math.floor(resp.status / 100) !== 2) {
+      throw new Error(`${resp.status} ${resp.statusText}: ${url}`);
+    }
+    return resp.json();
+  });
+}
+
+type ProgressFn = (more: any[]) => any;
+
+async function callPicasaP(p: number, paths: string[], progFn: ?ProgressFn = null): Promise<any[]> {
```
Comment (kalman) [inline samples/js/picasa/slurp/src/main.js]: done

File: samples/js/picasa/slurp/src/main.js (hunk 1 above)
Comment (kalman) [inline samples/js/picasa/slurp/src/main.js]: It doesn't appear to be.

File: samples/js/picasa/slurp/src/main.js (hunk 6 above)
Comment (aboodman) [inline samples/js/picasa/slurp/src/main.js]: yup

**Pull request merged**